crates/bevy_core/src/time/timer.rs

```diff
@@ -4,39 +4,56 @@ use bevy_property::Properties;
 use std::time::Duration;
 
 /// Tracks elapsed time. Enters the finished state once `duration` is reached.
+///
+/// Non repeating timers will stop tracking and stay in the finished state until reset.
+/// Repeating timers will only be in the finished state on each tick `duration` is reached or exceeded, and can still be reset at any given point.
 #[derive(Clone, Debug, Default, Properties)]
 pub struct Timer {
     pub elapsed: f32,
     pub duration: f32,
     pub finished: bool,
+    prev_finished: bool,
+    pub repeating: bool,
 }
 
 impl Timer {
-    pub fn new(duration: Duration) -> Self {
+    pub fn new(duration: Duration, repeating: bool) -> Self {
         Timer {
             duration: duration.as_secs_f32(),
+            repeating,
             ..Default::default()
         }
     }
 
-    pub fn from_seconds(seconds: f32) -> Self {
+    pub fn from_seconds(seconds: f32, repeating: bool) -> Self {
         Timer {
             duration: seconds,
+            repeating,
             ..Default::default()
         }
     }
 
     pub fn tick(&mut self, delta: f32) {
-        self.elapsed = (self.elapsed + delta).min(self.duration);
-        if self.elapsed >= self.duration {
-            self.finished = true;
+        if self.elapsed < self.duration {
+            self.elapsed += delta;
+        }
+
+        self.prev_finished = self.finished;
+        self.finished = self.elapsed >= self.duration;
+
+        if self.repeating && self.finished {
+            self.elapsed = self.elapsed % self.duration;
         }
     }
 
     pub fn reset(&mut self) {
         self.finished = false;
         self.elapsed = 0.0;
     }
+
+    pub fn just_finished(&self) -> bool {
+        self.prev_finished == false && self.finished == true
+    }
```

Comment:
This function is unnecessary in my opinion. It also does not work with `repeating` flag when the timer duration is smaller than the frame.

Reply:
Unless I'm overseeing simpler way to prevent a non-repeating finished timer from firing code every update tick, I think this saves a lot of repeated code when using non-repeating timers.

The timer not working with increments smaller than frametime seems like a different issue altogether, no?

Reply:
Just tested it. `just_finished` only fires once with 1 millisecond timer when `repeating` is true. It's happening because it finishes every frame, so `prev_finished` is always true.

Reply:
As for it's unnecessary. I would just put some logic around the timer in the system that makes it possible. But it is true that this way is simpler for that use case. So let's just ignore that.

Reply:
> Just tested it. just_finished only fires once with 1 millisecond timer when repeating is true.

Ah, I misunderstoord. I'll have another look at if there's a sensible way to handle this.

Reply:
So since repeating timers manage the `finished` state themselves, anytime `finished` is true should mean the timer just expired anyways, meaning it's safe to ignore `prev_finished` for repeating timers.